File: backend/app/scripts/songcache_reader.py
import os
import re
import json
import logging
from typing import List

logging.basicConfig(level=logging.DEBUG, format="%(asctime)s - %(levelname)s - %(message)s")

def clean_up_songs_data(songs):
    md5_groups = {}
    cleaned_songs = []
    deleted_songs = []

    # Group songs by MD5
    for song in songs:
        md5 = song["md5"]
        if md5 not in md5_groups:
            md5_groups[md5] = []
        md5_groups[md5].append(song)

    # Process each MD5 group
    for md5, group in md5_groups.items():
        if len(group) == 1 or not any(song["is_duplicate"] for song in group):
            cleaned_songs.extend(group)
        else:
            # Find the song with the longest name
            longest_name_song = max(group, key=lambda x: len(x["name"]))
            cleaned_songs.append(longest_name_song)
            logging.info(f"Longest name song: {longest_name_song['name']} - {longest_name_song['artist']} (MD5: {longest_name_song['md5']})")
            
            # Log deletions
            for song in group:
                if song != longest_name_song:
                    deleted_songs.append({
                        "md5": song["md5"],
                        "name": song["name"],
                        "artist": song["artist"],
                        "album": song["album"]
                    })
                    logging.info(f"Deleted duplicate song: {song['name']} - {song['artist']} (MD5: {song['md5']})")

    logging.info(f"Cleaned up {len(deleted_songs)} duplicate songs")
    return cleaned_songs, deleted_songs

def strip_color_tags(text: str) -> str:
    return re.sub(r'<color=[^>]+>(.*?)</color>', r'\1', text)

def strip_tags(text: str) -> str:
    return re.sub(r'</?b>', '', text)

def split_charters(charter_string: str) -> List[str]:
    char_delimiters = [",", "/", "&"]
    multichar_delimiters = [" - ", " + ", " and "]

    def is_in_color_tag(pos, text):
        open_tags = [m.start() for m in re.finditer(r'<color=', text[:pos])]
        close_tags = [m.start() for m in re.finditer(r'color>', text[:pos])]
        return len(open_tags) > len(close_tags)

    result = []
    current = []
    i = 0
    while i < len(charter_string):
        if any(charter_string.startswith(delim, i) for delim in multichar_delimiters) and not is_in_color_tag(i, charter_string):
            if current:
                result.append("".join(current).strip())
                current = []
            i += len(next(delim for delim in multichar_delimiters if charter_string.startswith(delim, i)))
        elif charter_string[i] in char_delimiters and not is_in_color_tag(i, charter_string):
            if current:
                result.append("".join(current).strip())
                current = []
            i += 1
        else:
            current.append(charter_string[i])
            i += 1

    if current:
        result.append("".join(current).strip())

    return [charter.strip() for charter in result if charter.strip()]

def process_charter_name(charter: str) -> str:
    return strip_tags(strip_color_tags(charter.strip())).strip()

def map_song_paths(base_path):
    """
    Map all valid song paths by searching for song.ini files.
    """
    song_paths = []
    for root, dirs, files in os.walk(base_path):
        if "song.ini" in files:
            song_paths.append(os.path.normpath(root).lower().replace("/", "\\"))
    logging.info(f"Found {len(song_paths)} song paths")
    return song_paths

def is_valid_md5(hash_string):
    """
    Check if a string is a valid MD5 hash.
    """
    return re.match(r"^[a-fA-F0-9]{32}$", hash_string) is not None

def find_md5_for_path(file_content, path):
    """
    Searches for the MD5 hash associated with the specified path in the binary file.
    """
    path_bytes = path.lower().encode("utf-8")
    start = 0
    md5_hashes = []

    while True:
        index = file_content.find(path_bytes, start)
        if index == -1:
            break

        # Find the start of the next path
        next_path_start = file_content.find(b"c:\\users\\jason\\documents\\clone hero\\songs", index + len(path_bytes))
        if next_path_start == -1:
            next_path_start = len(file_content)

        # Extract the data chunk between the current path and the next path
        data_chunk = file_content[index + len(path_bytes):next_path_start]
        
        # Extract the MD5 hash (first 32 characters of the last 36 characters)
        if len(data_chunk) >= 36:
            if data_chunk[-18] == 0:
                md5_hex = data_chunk[-17:-1].hex()
            else:
                md5_hex = data_chunk[-18:-2].hex()
            
            if is_valid_md5(md5_hex):
                md5_hashes.append(md5_hex)
                logging.info(f"Possible MD5 hash: {md5_hex}\nFor path: {path}")

        start = index + len(path_bytes)

    if not md5_hashes:
        logging.warning(f"No valid MD5 hash found for path: {path}")
    
    return md5_hashes

def process_songs(base_path, binary_file_path, output_file):
    """
    Process songs by mapping paths, finding MD5 hashes, and extracting song.ini data.
    """
    song_paths = map_song_paths(base_path)
    
    with open(binary_file_path, "rb") as bin_file:
        bin_file_content = bin_file.read()

    songs = []
    for path in song_paths:
        md5_hashes = find_md5_for_path(bin_file_content, path)
        if md5_hashes:
            ini_path = os.path.join(path, "song.ini")
            song_data = parse_ini_file(ini_path)

            charter_string = song_data.get("charter", "")
            charters = split_charters(charter_string)
            charter_refs = [process_charter_name(charter) for charter in charters]
            
            for md5_hex in md5_hashes:
                duplicate_songs = [song for song in songs if song["md5"] == md5_hex]
                if duplicate_songs:
                    is_duplicate = True
                    if duplicate_songs[0]["artist"] == song_data.get("artist", "") and duplicate_songs[0]["name"] == song_data.get("name", "") and duplicate_songs[0]["album"] == song_data.get("album", "") and duplicate_songs[0]["track"] == song_data.get("track", ""):
                        continue
                else:
                    is_duplicate = False
                song_info = {
                    "md5": md5_hex,
                    "artist": song_data.get("artist", ""),
                    "name": song_data.get("name", ""),
                    "album": song_data.get("album", ""),
                    "track": song_data.get("track", ""),
                    "year": song_data.get("year", ""),
                    "genre": song_data.get("genre", ""),
                    "difficulty": song_data.get("diff_drums", ""),
                    "song_length": song_data.get("song_length", ""),
                    "charter_refs": ",".join(charter_refs),
                    "is_duplicate": is_duplicate
                }
                songs.append(song_info)
        else:
            logging.warning(f"MD5 not found for path: {path}")

    cleaned_songs, deleted_songs = clean_up_songs_data(songs)

    with open(output_file, "w", encoding="utf-8") as file:
        json.dump(cleaned_songs, file, indent=2)
    logging.info(f"Processed data saved to {output_file}.")

    deleted_log_file = output_file.replace(".json", "_deleted_log.json")
    with open(deleted_log_file, "w", encoding="utf-8") as file:
        json.dump(deleted_songs, file, indent=2)
    logging.info(f"Deleted songs log saved to {deleted_log_file}.")

def parse_ini_file(ini_path):
    """
    Parse the song.ini file to extract required fields, attempting to handle different encodings.
    """
    data = {}
    required_fields = [
        "artist", "name", "album", "track", "year",
        "genre", "diff_drums", "song_length", "charter"
    ]
    try:
        with open(ini_path, "r", encoding="utf-8") as file:
            lines = file.readlines()
    except UnicodeDecodeError:
        try:
            with open(ini_path, "r", encoding="cp1252") as file:
                lines = file.readlines()
        except UnicodeDecodeError:
            logging.error(f"Failed to decode {ini_path}. Skipping.")
            return data

    for line in lines:
        key, _, value = line.partition("=")
        key = key.strip().lower()
        value = value.strip()
        if key in required_fields:
            data[key] = value
    return data

def format_song_length(milliseconds):
    """
    Convert song length from milliseconds to HH:MM:SS format.
    """
    seconds = int(milliseconds) // 1000
    minutes, seconds = divmod(seconds, 60)
    hours, minutes = divmod(minutes, 60)
    return f"{hours:02}:{minutes:02}:{seconds:02}"

if __name__ == "__main__":
    base_path = "C:\\Users\\jason\\Documents\\Clone Hero\\Songs"
    binary_file_path = "C:\\Users\\jason\\AppData\\LocalLow\\srylain Inc_\Clone Hero\\songcache.bin"
    output_file = "data\\songs_with_md5.json"
    
    process_songs(base_path, binary_file_path, output_file)

File: src/scripts/songcache_reader.ts
import * as fs from "fs";
import * as path from "path";
import { scanChartFolder, ScannedChart } from "scan-chart";

interface ScannedChartExtra extends ScannedChart {
    playlistPath?: string;
}

const logging = console;

function cleanUpSongsData(songs: ScannedChartExtra[]): { cleanedSongs: ScannedChartExtra[], deletedSongs: ScannedChartExtra[] } {
    const md5Groups: { [key: string]: ScannedChartExtra[] } = {};
    const cleanedSongs: ScannedChartExtra[] = [];
    const deletedSongs: ScannedChartExtra[] = [];

    // Group songs by MD5
    for (const song of songs) {
        if (!md5Groups[song.md5]) {
            md5Groups[song.md5] = [];
        }
        md5Groups[song.md5].push(song);
    }

    // Process each MD5 group
    for (const group of Object.values(md5Groups)) {
        if (group.length === 1) {
            cleanedSongs.push(group[0]);
        } else {
            // Sort songs by name length (descending)
            const sortedSongs = [...group].sort((a, b) => 
                ((b.name?.length || 0) - (a.name?.length || 0))
            );
            
            // Keep track of songs we've decided to keep
            const songsToKeep = new Set<ScannedChartExtra>();
            
            // For each song, check if it's a distinct song or a variant of one we've seen
            for (let i = 0; i < sortedSongs.length; i++) {
                const currentSong = sortedSongs[i];
                let isDuplicate = false;
                
                // Check against songs we've already decided to keep
                for (const keptSong of songsToKeep) {
                    // If one name is a substring of the other, consider it a duplicate
                    const currentName = (currentSong.name || "").toLowerCase();
                    const keptName = (keptSong.name || "").toLowerCase();
                    
                    // Don't consider it a duplicate if they're exactly the same name
                    // This handles the case where two songs with identical names might be different songs
                    if (currentName !== keptName && 
                       (currentName.includes(keptName) || keptName.includes(currentName))) {
                        isDuplicate = true;
                        deletedSongs.push(currentSong);
                        break;
                    }
                }
                
                if (!isDuplicate) {
                    songsToKeep.add(currentSong);
                    cleanedSongs.push(currentSong);
                }
            }
            
            logging.info(`Found ${songsToKeep.size} distinct songs in group of ${group.length} with same MD5`);
        }
    }

    logging.info(`Cleaned up ${deletedSongs.length} duplicate songs`);
    return { cleanedSongs, deletedSongs };
}

function splitCharters(charterString: string): string[] {
    const charDelimiters = [",", "/", "&"];
    const multicharDelimiters = [" - ", " + ", " and "];

    const result: string[] = [];
    let current: string[] = [];
    let i = 0;

    while (i < charterString.length) {
        if (multicharDelimiters.some(delim => charterString.startsWith(delim, i))) {
            if (current.length) {
                result.push(current.join("").trim());
                current = [];
            }
            i += multicharDelimiters.find(delim => charterString.startsWith(delim, i))!.length;
        } else if (charDelimiters.includes(charterString[i])) {
            if (current.length) {
                result.push(current.join("").trim());
                current = [];
            }
            i++;
        } else {
            current.push(charterString[i]);
            i++;
        }
    }

    if (current.length) {
        result.push(current.join("").trim());
    }

    return result.filter(charter => charter.trim() !== "");
}

function mapSongPaths(basePath: string): string[] {
    const songPaths: string[] = [];
    let processedFolders = 0;
    const maxPathLength = process.stdout.columns - 40;

    const walkSync = (dir: string) => {
        const files = fs.readdirSync(dir);
        processedFolders++;
        let displayDir = dir;
        if (dir.length > maxPathLength) {
            displayDir = "..." + dir.slice(-maxPathLength);
        }
        process.stdout.write(`\rWalking through folder ${processedFolders}: ${displayDir}`);
        for (const file of files) {
            const filePath = path.join(dir, file);
            const stat = fs.statSync(filePath);
            if (stat.isDirectory()) {
                walkSync(filePath);
            } else if (file.toLowerCase() === "song.ini") {
                songPaths.push(path.dirname(filePath).toLowerCase().replace(/\//g, "\\"));
            }
        }
    };
    walkSync(basePath);
    process.stdout.write("\n");
    logging.info(`Found ${songPaths.length} song paths`);
    return songPaths;
}

function isValidMd5(hash: string): boolean {
    return /^[a-fA-F0-9]{32}$/.test(hash);
}

// findMd5ForPath with optional skipMd5 parameter to avoid specific MD5
function findMd5ForPath(fileContent: Buffer, path: string, skipMd5s: Set<string> = new Set()): string | null {
    const pathBytes = Buffer.from(path.toLowerCase());
    let start = 0;
    let md5Hash: string | null = null;

    while (true) {
        const index = fileContent.indexOf(pathBytes, start);
        if (index === -1) break;

        const nextPathStart = fileContent.indexOf(Buffer.from("d:\\songs"), index + pathBytes.length);
        const endIndex = nextPathStart === -1 ? fileContent.length : nextPathStart;

        const dataChunk = fileContent.subarray(index + pathBytes.length, endIndex);

        if (dataChunk.length >= 36) {
            let md5Hex: string;
            if (dataChunk[dataChunk.length - 18] === 0 || dataChunk[dataChunk.length - 18] === 107) {
                md5Hex = dataChunk.subarray(-17, -1).toString("hex");
            } else {
                md5Hex = dataChunk.subarray(-18, -2).toString("hex");
            }

            if (isValidMd5(md5Hex) && !skipMd5s.has(md5Hex)) {
                md5Hash = md5Hex;
                break;
            }
        }

        start = index + pathBytes.length;
    }

    if (!md5Hash) {
        logging.warn(`No valid MD5 hash found for path: ${path}${skipMd5s.size > 0 ? ' (skipped ' + skipMd5s.size + ' existing MD5s)' : ''}`);
    }

    return md5Hash;
}

async function processSongs(basePath: string, binaryFilePath: string, outputFile: string): Promise<void> {
    const songPaths = mapSongPaths(basePath);
    const songs: ScannedChartExtra[] = [];
    const maxPathLength = process.stdout.columns - 40;
    const binaryFileContent = fs.readFileSync(binaryFilePath);
    const assignedMd5s = new Map<string, string>(); // Map from path to MD5
    
    // Group song paths by parent directory
    const songPathsByParent: {[key: string]: string[]} = {};
    for (const path of songPaths) {
        const parentDir = path.split("\\").slice(0, -1).join("\\");
        if (!songPathsByParent[parentDir]) {
            songPathsByParent[parentDir] = [];
        }
        songPathsByParent[parentDir].push(path);
    }
    
    // Process songs by directory, sorting paths by length (longest first)
    let processedSongs = 0;
    const totalSongs = songPaths.length;
    
    for (const parentDir of Object.keys(songPathsByParent)) {
        // Sort paths by length (longest first) to process more specific paths first
        const sortedPaths = songPathsByParent[parentDir].sort((a, b) => b.length - a.length);
        const dirMd5s = new Set<string>(); // Track MD5s used in this directory
        
        for (const path of sortedPaths) {
            processedSongs++;
            let displayDir = path;
            if (path.length > maxPathLength) {
                displayDir = "..." + path.slice(-maxPathLength);
            }
            process.stdout.write(`\rProcessing song ${processedSongs} of ${totalSongs}: ${displayDir}`);
            
            try {
                const files = fs.readdirSync(path).map(file => ({
                    fileName: file,
                    data: fs.readFileSync(path + "/" + file)
                }));
                const scannedChart = scanChartFolder(files);
                const playlistPath = path
                    .replace("d:\\songs\\", "")
                    .replace("downloaded songs\\", "")
                    .replace("rclone songs\\sync charts\\", "");
                const scannedChartExtra = { ...scannedChart, playlistPath: playlistPath };
                scannedChartExtra.albumArt = null;
                if (scannedChartExtra.notesData) {
                    scannedChartExtra.notesData.chartIssues = [];
                }
                if (scannedChartExtra.charter) {
                    scannedChartExtra.charter = splitCharters(scannedChartExtra.charter).join(",");
                }
                
                // Find MD5, skipping any that are already used in this directory
                // This prevents incorrect assignment when similar paths exist
                const md5 = findMd5ForPath(binaryFileContent, path, dirMd5s);
                if (md5) {
                    assignedMd5s.set(path, md5);
                    dirMd5s.add(md5); // Mark this MD5 as used in this directory
                    scannedChartExtra.md5 = md5;
                    songs.push(scannedChartExtra);
                } else {
                    logging.error(`Failed to find MD5 for path: ${path}`);
                }
            } catch (error) {
                logging.error(`Error processing path: ${path}`, error);
            }
        }
    }
    
    process.stdout.write("\n");
    const { cleanedSongs, deletedSongs } = cleanUpSongsData(songs);

    fs.writeFileSync(outputFile, JSON.stringify(cleanedSongs, null, 2));
    logging.info(`Processed data saved to ${outputFile}.`);

    const deletedLogFile = outputFile.replace(".json", "_deleted_log.json");
    fs.writeFileSync(deletedLogFile, JSON.stringify(deletedSongs, null, 2));
    logging.info(`Deleted songs log saved to ${deletedLogFile}.`);
}

const basePath = "D:\\Songs";
const binaryFilePath = "C:\\Users\\jason\\AppData\\LocalLow\\srylain Inc_\\Clone Hero\\songcache.bin";
const outputFile = "data\\songs_with_md5.json";

processSongs(basePath, binaryFilePath, outputFile).catch(error => {
    console.error("Error processing songs:", error);
});